### PR TITLE
[Refactor][Conv1D] align conv1d kernels and ops with spec

### LINKED ISSUE #853: [REFACTOR][CONV1D] remediate conv1d_fwd manifest spec-only gaps
### Symptom / Motivation

`conv1d_fwd` and `conv1d_fwd_bias` manifest entries are `spec-only` because the current op/kernel implementation diverges from the PyTorch `torch.nn.functional.conv1d` API in several ways. The manifest was added to align with PyTorch's public interface (R15), but the code does not yet conform.

### Root Cause Analysis

Semantic gap analysis from `.foundry/migrations/conv.json`:

**1. Parameter naming (`trivial`)**
- `Conv1dOp.forward()` uses `x` for the first input; PyTorch (and manifest) uses `input`.
- Rename only — no logic change.

**2. `dilation` param (`kernel-change`)**
- PyTorch supports `dilation` (default 1). The kernel (`tileops/kernels/conv/conv1d.py`) computes `L_out` without dilation: `out_l = (l_in + 2 * pad_l - kernel_l) // stride_l + 1`.
- The im2col-style unrolling in the pipelined loop (lines 64-83) indexes `il = ol * stride_l + kw - pad_l`. Dilation requires changing this to `il = ol * stride_l + kw * dilation - pad_l` and adjusting `L_out = (L_in + 2*padding - dilation*(kW-1) - 1) // stride + 1`.
- Kernel architectural change needed: the `k_total` computation and weight flattening remain the same (`k_total = kernel_l * c_in`), but the spatial index mapping changes.

**3. `groups` param (`kernel-change`)**
- PyTorch supports grouped convolution (`groups` default 1). The kernel assumes `weight.shape = (kernel_l, c_in, c_out)` with full `c_in`. Grouped conv requires weight shape `(kernel_l, c_in // groups, c_out)` and restricting each group's input channels.
- The kernel's im2col approach flattens `(kw, ci)` into `k_total`. For groups, the GEMM must be partitioned: each group computes `(N * L_out, C_out // groups)` from `(N * L_out, kW * C_in // groups)`. This requires either looping over groups or restructuring the kernel grid.
- Significant architectural rework.

**4. `padding` as `str` (`trivial`)**
- PyTorch accepts `padding='valid'` (0) and `padding='same'`. The manifest declares `type: "int | str"`. The op's `__init__` only accepts `int`. String padding is a convenience that maps to computed int values — can be handled in the Op layer without kernel changes.

**5. Variant split alignment (`trivial`)**
- `Conv1dOp.forward()` currently accepts `Optional[bias]`. The manifest splits into `conv1d_fwd` (no bias) and `conv1d_fwd_bias` (required bias). The op needs to be split into two classes (or a base + variant) per R16 codegen rules.

**6. Class structure**
- `Conv1dOp` inherits from `Op` directly. The variant split (R16) requires introducing either a base class or delegation pattern. Since `conv2d` and `conv3d` exist in the same family, a shared `_ConvBaseOp` may be warranted — but that decision affects the broader conv family and should be evaluated holistically.

**7. No cascading dependencies**
- `conv2d` and `conv3d` have analogous gaps (same missing params). Fixing conv1d first establishes the pattern for the family.

### Related Files

- `tileops/ops/conv1d.py`
- `tileops/kernels/conv/conv1d.py`
- `tests/ops/test_conv1d.py`
- `benchmarks/ops/bench_conv1d.py`
- `tileops/ops_manifest.yaml`
- `.foundry/migrations/conv.json`

## Goal

Bring `conv1d_fwd` and `conv1d_fwd_bias` to `status: implemented` by aligning the op and kernel code with the manifest spec (PyTorch API).

## Plan

1. **Rename `x` → `input`** in `Conv1dOp.forward()` signature (trivial)
2. **Add string padding support** in `Conv1dOp.__init__()` — resolve `'valid'`/`'same'` to int before passing to kernel (trivial, op-layer only)
3. **Add `dilation` to kernel** — modify spatial index `il = ol * stride_l + kw * dilation - pad_l` and `L_out` formula in `_conv1d_kernel` (kernel-change)
4. **Add `groups` to kernel** — partition the GEMM by group, adjust weight shape handling (kernel-change, most effort)
5. **Split `Conv1dOp` into variant classes** per R16 — `Conv1dFwdOp` (no bias) and `Conv1dFwdBiasOp` (required bias), sharing kernel dispatch logic
6. **Update tests** to cover `dilation > 1`, `groups > 1`, string padding, and the no-bias variant
7. **Run `validate_manifest.py --check-op conv1d_fwd`** — confirm L0–L4 pass, flip status to `implemented`

## Constraints

- Do not modify `ops_manifest.yaml` in this issue's PR — manifest changes go in their own PR (trust model).
- Kernel changes must not regress existing conv1d benchmarks.
- Do not address conv2d/conv3d in this issue — they follow as separate issues once the conv1d pattern is established.

## Acceptance Criteria

- [ ] AC-1: Modified files pass existing tests (`pytest tests/ops/test_conv1d.py`)
- [ ] AC-2: `python scripts/validate_manifest.py --check-op conv1d_fwd` passes L0–L4 (signature, shape, dtype, bench)
- [ ] AC-3: `Conv1dOp.forward()` uses `input` (not `x`) as first param name
- [ ] AC-4: Kernel supports `dilation` parameter (verified by test with `dilation=2`)
- [ ] AC-5: Kernel supports `groups` parameter (verified by test with `groups > 1`)
- [ ] AC-6: Op accepts `padding='valid'` and `padding='same'` string values
- [ ] AC-7: Two distinct Op classes exist for the no-bias and bias variants

### PR DESCRIPTION
**Pull request by RMLYC**
## Summary

Align Conv1d ops and kernels with the manifest/PyTorch-facing spec tracked in #853.

## What changed

- add `Conv1dFwdOp` and `Conv1dFwdBiasOp` variants and remove the legacy `Conv1dOp` bridge
- align the op interface with PyTorch-style `input`, `padding`, `dilation`, and `groups`
- support `padding="valid"` and `padding="same"` without extra `F.pad` preprocessing
- add a grouped conv1d kernel path alongside the single-group kernel path under `Conv1dKernel`
- move Conv1d parameter/layout/tensor validation into shared conv common helpers
- update Conv1d tests and benchmarks to cover dilation, grouped convolution, string padding, and no-bias variants

## Validation

- `python -m pytest tests/ops/test_conv1d.py -v`
- `python -m pytest benchmarks/ops/bench_conv1d.py -v -k 'seanet-k3-same-fp16'`

## Notes

- `python scripts/validate_manifest.py --check-op conv1d_fwd` is still blocked on this branch because `conv1d_fwd` is not present in the current `ops_manifest.yaml`.

## Refactor Summary

FAMILY: conv

PROMOTED: 0
- none

BLOCKED: 2
- `conv1d_fwd`: code, tests, and benchmark updates are in place, but `python scripts/validate_manifest.py --check-op conv1d_fwd` is currently blocked because `conv1d_fwd` is not present in the current `tileops/ops_manifest.yaml`
- `conv1d_fwd_bias`: blocked for the same reason; the bias variant path is implemented, but there is no manifest entry on this branch to validate or promote

CLEANUP_GROUPS: conv1d
- completed cleanup for the conv1d group by removing the legacy `Conv1dOp` bridge and keeping only explicit `Conv1dFwdOp` / `Conv1dFwdBiasOp` variants
- `Conv1dKernel` now distinguishes single-group and grouped kernel builder paths under one wrapper class
- Conv1d parameter, layout, and tensor validation has been moved into shared conv common helpers

FOLLOW_UPS:
- add or restore `conv1d_fwd` and `conv1d_fwd_bias` entries in `tileops/ops_manifest.yaml`
- rerun:
  - `python scripts/validate_manifest.py --check-op conv1d_fwd`
  - `python scripts/validate_manifest.py --check-op conv1d_fwd_bias`
  - `python -m pytest tests/ops/test_conv1d.py -v`
- once manifest validation passes, perform manifest status promotion only in the orchestrator stage

Validation completed on this branch:
- `python -m pytest tests/ops/test_conv1d.py -v` passed
- `python -m pytest benchmarks/ops/bench_conv1d.py -v -k 'seanet-k3-same-fp16'` passed

Closes #853

### CODE REVIEW

File: tileops/kernels/conv/conv1d.py [hunk 1]
```diff
@@ -157,10 +284,77 @@ def _(
...
+    dtype: str,
+    block_m: int,
+    block_n: int,
+    block_k: int,
+    num_stages: int,
+    threads: int,
+    enable_rasterization: bool,
+    *inputs: tuple[torch.Tensor, ...],
+) -> torch.Tensor:
+    _ = (groups, c_in_per_group, has_bias, dtype, block_m, block_n, block_k, num_stages, threads, enable_rasterization)
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tileops/kernels/conv/conv1d.py]: Updated in `00c17a3`. I removed that unused tuple assignment from the fake registration path instead of reformatting it.

File: tileops/kernels/conv/conv1d.py [hunk 2]
```diff
@@ -112,6 +125,116 @@ def _conv1d_main(
...
+                        k_idx = k_iter * block_k + i
+                        oc = bx * block_n + j
+                        kw = k_idx // c_in_per_group
+                        ci = k_idx % c_in_per_group
+                        weight_shared[i, j] = T.if_then_else(
+                            (k_idx < k_total) & (oc < c_out_per_group),
+                            weight[bz, kw, ci, oc],
+                            T.cast(0.0, dtype),
+                        )
+                    T.gemm(data_shared, weight_shared, out_local)
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tileops/kernels/conv/conv1d.py]: Updated in `00c17a3`. The grouped conv1d path now materializes a flattened grouped weight view and uses `T.copy(...)` into `weight_shared` instead of scalar per-element shared-memory loads.

**Comment by lcy-seso** [comment]
Hi, just checking in — do you plan to make any further updates to this PR? If not, I’ll close it for now. Please feel free to reopen it anytime if you’d like to continue working on it.
